solr/test-framework/src/java/org/apache/solr/SolrTestCaseJ4.java

```diff
@@ -2655,14 +2656,25 @@ public static HttpClient getHttpClient(String url) {
   /**
    * This method creates a basic HttpSolrClient. Tests that want to control the creation process
    * should use the {@link org.apache.solr.client.solrj.impl.Http2SolrClient.Builder} class directly
+   *
+   * @param url the base URL for a Solr node. Should not contain a core or collection name.
    */
   public static HttpSolrClient getHttpSolrClient(String url) {
     return new HttpSolrClient.Builder(url).build();
   }
 
+  /** Create a basic HttpSolrClient pointed at the specified replica */
+  public static HttpSolrClient getHttpSolrClient(Replica replica) {
+    return getHttpSolrClient(replica.getBaseUrl(), replica.getCoreName());
+  }
+
   /**
    * This method creates a basic HttpSolrClient. Tests that want to control the creation process
    * should use the {@link org.apache.solr.client.solrj.impl.Http2SolrClient.Builder} class directly
+   *
+   * @param url the base URL of a Solr node. Should <em>not</em> include a collection or core name.
+   * @param defaultCoreName the name of a core that the created client should default to when making
+   *     core-aware requests
    */
   public static HttpSolrClient getHttpSolrClient(String url, String defaultCoreName) {
```

Comment:
I wonder if a check for the situation where the defaultCoreName is in the url?   and throw an exception?

Reply:
On `branch_9x`, afaict, there's no correct way to detect when a URL contains a core name.  I actually caused a bunch of test failures [earlier this week](https://issues.apache.org/jira/browse/SOLR-17066?focusedCommentId=17814783&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17814783) by not realizing this.  The root of the problem is that 9.x releases allow users to use arbitrary hostContext's or context-root's or whatever - in such a world there's no way to look at the URL and determine a collection name from anything else.

That said - on `main` where the context-root is "locked in" I 100% agree we should have some check for this.  But IMO it should go in the client ctors directly, so that it gets triggered in **all** cases.